src/java/org/apache/cassandra/index/sai/StorageAttachedIndexBuilder.java

```diff
@@ -324,7 +324,7 @@ private void completeSSTable(SSTableFlushObserver indexWriter,
         }
 
         // register custom index components into existing sstables
-        sstable.registerComponents(group.getComponents(existing), tracker);
+        sstable.registerComponents(group.getLiveComponents(sstable, existing), tracker);
```

Comment:
could this change stop the "incomplete" check from working?  I think that is supposed to force index build during startup if an index does not exist, but we are not ignoring components that don't exist?

Reply:
`completeSSTable` is only called when an index has built successfully and the completion markers have been written. The issue here is that we were registering components that didn't exist on disk.

Reply:
sure, but if we only register the components that exist on disk how can the incomplete list find "missing" components.  If I dig down into `group.onSSTableChanged` below I end up in `org.apache.cassandra.index.sai.SSTableContextManager#update` which seems to be using the registered components to verify all the components exist?

Reply:
Ok, digging deeper, down in `org.apache.cassandra.index.sai.ColumnContext#getBuiltIndexes` it checks if an index was not built because there was no data.  So I think this should be OK.

Reply:
Ok, before I merge this one, I'm going to go right through it again but I'm pretty certain this is correct.

Reply:
I've looked at this again.

The registering of components with the SSTable and the management of index queryability are tangential. 

At the point where we are registering the components we are telling the SSTable that we have created these on-disk components and that you (the SSTable) are now responsible for backing them up, streaming them, deleting them. We are not making any assertions to the SSTable that the index is queryable. 

Queryability of the index is handled by the `StorageAttachedIndexGroup.onSSTableChanged` which after calling `SSTableContextManager.update` will delete any on-disk components if the index is invalid and mark the index as not queryable. This last bit could be a problem except that, because this is a newly created index, we do not validate the index components during the update so they cannot be invalidated and deleted.

I think we are safe with the ordering of these methods so am going to merge this PR.

Reply:
Thanks for double checking it.